src/main/java/io/divolte/server/recordmapping/Digester.java

```diff
@@ -0,0 +1,193 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.divolte.server.recordmapping;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.reflect.TypeToken;
+import io.divolte.server.DivolteEvent;
+import io.divolte.server.recordmapping.DslRecordMapping.ValueProducer;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+@ParametersAreNonnullByDefault
+public final class Digester<T> {
+    private final String digestIdentifier;
+    private final Supplier<T> digestFactory;
+    private final Function<T,Consumer<ByteBuffer>> digestUpdaterFactory;
+    private final Function<T,Supplier<byte[]>> digestFinalizerFactory;
+    private List<ValueProducer<ByteBuffer>> digestPieces = new ArrayList<>();
+
+    private Digester(final String identifier,
+                     final Supplier<T> digestFactory,
+                     final Function<T,Consumer<ByteBuffer>> digestUpdater,
+                     final Function<T,Supplier<byte[]>> digestFinalizer) {
+        digestIdentifier = Objects.requireNonNull(identifier);
+        this.digestFactory = Objects.requireNonNull(digestFactory);
+        this.digestUpdaterFactory = Objects.requireNonNull(digestUpdater);
+        this.digestFinalizerFactory = Objects.requireNonNull(digestFinalizer);
+    }
+
+    public ValueProducer<ByteBuffer> result() {
+        final String identifier =
+            digestPieces.stream()
+                        .map(s -> String.format(".add(%s)", s))
+                        .collect(Collectors.joining("", digestIdentifier, ".result()"));
+        return new BytesValueProducer(identifier, this::calculateDigest);
+    }
+
+    private Optional<ByteBuffer> calculateDigest(final DivolteEvent e, final Map<String,Optional<?>> context) {
+        final T messageDigest = digestFactory.get();
+        final Consumer<ByteBuffer> digestUpdater = digestUpdaterFactory.apply(messageDigest);
+        final Supplier<byte[]> digestFinalizer = digestFinalizerFactory.apply(messageDigest);
+        digestPieces.stream()
+                    .map(x -> x.produce(e, context))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
```

Comment:
I'm not doing so much Java lately, but no flapMap available?

Reply:
There is `flatMap()` but the function has to return a `Stream` and Java 8 lacks an easy way to convert `Optional` into a `Stream`. (In Java 9 this is fixed.)

However I just noticed Guava has a helper so that's what we'll do.